### PR TITLE
PID API: fixed error with struct and cleaned up utility functions

### PR DESCRIPTION
**Pull request by younesr1**
These changes only affect my work on the PID API. I decided it made more sense for the payload to stringify itself that what i previously had. I also realise a dumb mistake with the struct declaration (declared instance and not a type). One thing i'm worried about is that adding a function to the payload means the function pointer will be transmitted over CAN. I tried to verify this by creating a dummy program and it looks like function pointers are not included in the sizeof() of a struct so I dont think it will be transmitted. @jetkov @wmmc88  what do you think?

### CODE REVIEW

**Comment by wmmc88** [comment]
I don't think you should be serializng a struct with the funciton. Why send that function over every time when its the same definition. it should just be outside of the struct.

**Comment by younesr1** [comment]
> I don't think you should be serializng a struct with the funciton. Why send that function over every time when its the same definition. it should just be outside of the struct.

**Comment by younesr1** [comment]
> I don't think you should be serializng a struct with the funciton. Why send that function over every time when its the same definition. it should just be outside of the struc

i agree. now that i think abt it, rhe function ptr is def sent every time.

File: include/hw_bridge.h [hunk 1]
```diff
@@ -109,17 +109,15 @@ enum canID_t {
 namespace ARM {
 namespace ACTUATOR {
 enum actuatorID_t { TURNTABLE, SHOULDER, ELBOW, WRISTLEFT, WRISTRIGHT, CLAW };
-std::string stringifyActuatorID(actuatorID_t actuator);
 }  // namespace ACTUATOR
 namespace PID {
-typedef struct __attribute__((__packed__)) {
+struct __attribute__((__packed__)) tuningApiPayload {
```
Comment (younesr1) [inline include/hw_bridge.h]: compiler says typedef does nothing and throws an error

Comment (wmmc88) [reply include/hw_bridge.h]: Whats the error?

Comment (jetkov) [reply include/hw_bridge.h:114]: To map the typdef to something you need to add a name after the closing '}', without it it does nothing. Ex `typedef uint8_t byte` vs `typedef uint8_t`. The name before the opening '{' is a tag not a type.

Comment (younesr1) [reply include/hw_bridge.h]: i just misused typedef. alex explained it.

File: include/hw_bridge.h [hunk 2]
```diff
@@ -108,18 +108,16 @@ enum canID_t {
 
 namespace ARM {
 namespace ACTUATOR {
-enum actuatorID_t { TURNTABLE, SHOULDER, ELBOW, WRISTLEFT, WRISTRIGHT, CLAW };
-std::string stringifyActuatorID(actuatorID_t actuator);
+enum ActuatorId { TURNTABLE, SHOULDER, ELBOW, WRISTLEFT, WRISTRIGHT, CLAW };
 }  // namespace ACTUATOR
 namespace PID {
-typedef struct __attribute__((__packed__)) {
+using tuningApiPayload = struct __attribute__((__packed__)) tuningApiPayload {
```
Comment (younesr1) [inline include/hw_bridge.h]: CI running on sw complained abt this

File: include/hw_bridge.h [hunk 3]
```diff
@@ -108,18 +108,16 @@ enum canID_t {
...
   float value;
   bool velocity;
-  uint8_t actuatorID;
-} payload;
-enum parameter_t { P, I, D, DEADZONE, BIAS };
-std::string stringifyParam(parameter_t param);
-std::string stringifyVelPos(bool vel_pos);
+  HWBRIDGE::ARM::ACTUATOR::ActuatorId actuatorID;
+};
+std::string str(const tuningApiPayload& payload);
```
Comment (younesr1) [inline include/hw_bridge.h]: hmm useless to return a const string? im suprised: 
    return type 'const std::string' (aka 'const basic_string<char>') is 'const'-qualified at the top level, which may reduce code readability without improving const correctness [readability-const-return-type,-warnings-as-errors]

Comment (wmmc88) [reply include/hw_bridge.h]: the practical reason is that returning a const string will prevent its resources from being moved

Comment (wmmc88) [reply include/hw_bridge.h]: Another reason is that "why should you prevent modification" in the first place. If the dev wants a const string, they can `const std::string s = funciton_that_returns_string();`

Comment (wmmc88) [reply include/hw_bridge.h]: you can read more about const correctness here: https://isocpp.org/wiki/faq/const-correctness#overview-const

Comment (wmmc88) [reply include/hw_bridge.h]: also u gotta remember that `std::string` is dynamically allocated so making it const is not a good idea

Comment (wmmc88) [reply include/hw_bridge.h]: returning a const c string is usually fine though

Comment (wmmc88) [reply include/hw_bridge.h]: Is this referring to the correct line of code? or is it supposed to refer to `std::string HWBRIDGE::ARM::PID::str(const TuningApiPayload& payload) `

Comment (younesr1) [reply include/hw_bridge.h]: i agree about how returning a const string doesnt make sense. what do you mean by your last comment?

**Comment by younesr1** [comment]
This PR ended up containing some stuff needed for porting this repo to sw and also for my pid parameter tuning over can task on fw. i will break them up into 2 PRs

File: include/hw_bridge.h (hunk 3 above)
Comment (wmmc88) [reply include/hw_bridge.h]: which comment?

Comment (younesr1) [reply include/hw_bridge.h]: Is this referring to the correct line of code? or is it supposed to refer to std::string HWBRIDGE::ARM::PID::str(const TuningApiPayload& payload)

Comment (wmmc88) [reply include/hw_bridge.h]: This is the prototype for that function?

File: src/hw_bridge.cpp [hunk 4]
```diff
@@ -1,41 +1,35 @@
 #include "hw_bridge.h"
 
-std::string HWBRIDGE::ARM::ACTUATOR::stringifyActuatorID(HWBRIDGE::ARM::ACTUATOR::ActuatorId actuator) {
-  switch (actuator) {
+#include <iomanip>
+#include <sstream>
+
+std::string HWBRIDGE::ARM::PID::str(const TuningApiPayload& payload) {
```
Comment (wmmc88) [inline src/hw_bridge.cpp]: argument has different name from prototype in header

File: src/hw_bridge.cpp [hunk 5]
```diff
@@ -1,41 +1,35 @@
...
-  }
-}
-
-std::string HWBRIDGE::ARM::PID::stringifyVelPos(bool isVelocityPID) {
-  return isVelocityPID ? "velocity" : "position";
+  std::string control = payload.isVelocityPID ? "VELOCITY" : "POSITION";
+  std::stringstream strm;
+  strm << std::fixed << std::setprecision(4) << payload.value;
+  std::string val = strm.str();
+  return "{Value: " + val + ", Actuator: " + actuator + ", Control: " + control + "}";
```
Comment (wmmc88) [inline src/hw_bridge.cpp]: nit: imo val should be last

File: include/hw_bridge.h (hunk 3 above)
Comment (younesr1) [reply include/hw_bridge.h]: yeah

File: src/hw_bridge.cpp [hunk 6]
```diff
@@ -1,41 +1,35 @@
 #include "hw_bridge.h"
 
-std::string HWBRIDGE::ARM::ACTUATOR::stringifyActuatorID(HWBRIDGE::ARM::ACTUATOR::ActuatorId actuator) {
-  switch (actuator) {
+#include <iomanip>
+#include <sstream>
+
+std::string HWBRIDGE::ARM::PID::str(const TuningApiPayload& payload) {
+  std::string actuator;
```
Comment (wmmc88) [inline src/hw_bridge.cpp:11]: C++ 20 standard actually implemented constexpr std::string [but literally no compiler supports it yet LOL](https://en.cppreference.com/w/cpp/compiler_support). Next best thing i think here is to return a constexpr string_view.

Comment (wmmc88) [reply src/hw_bridge.cpp:11]: A somewhat similar example I just wrote up:
https://wandbox.org/permlink/9RQ9WcBzFibocAx5

File: include/hw_bridge.h (hunk 3 above)
Comment (wmmc88) [reply include/hw_bridge.h]: so its referring to the same thing?

File: include/hw_bridge.h [hunk 7]
```diff
@@ -109,17 +109,15 @@ enum CanId {
 namespace ARM {
 namespace ACTUATOR {
 enum ActuatorId { TURNTABLE, SHOULDER, ELBOW, WRISTLEFT, WRISTRIGHT, CLAW };
```
Comment (wmmc88) [inline include/hw_bridge.h]: always prefer to use enum class instead of enum

**Review by wmmc88 — CHANGES_REQUESTED** [review]

File: include/hw_bridge.h (hunk 3 above)
Comment (wmmc88) [reply include/hw_bridge.h]: see my recent comments on the implementation of this prototype.

File: include/hw_bridge.h (hunk 7 above)
Comment (younesr1) [reply include/hw_bridge.h]: ive applied that to the enums im using and created a task to change for all enums in this repo

Comment (wmmc88) [reply include/hw_bridge.h]: noice

File: src/hw_bridge.cpp (hunk 6 above)
Comment (wmmc88) [reply src/hw_bridge.cpp:11]: ?

**Pull request closed without merging**